### ISSUE TITLE
Should the SPG take interlacing into account?

### ISSUE THREAD

**Issue by snickerbockers** [washingtondc-emu/washingtondc#43]
For a while I've been observing that games which probably should be running at 60Hz are instead running at 30Hz.  This has also caused the game logic to run at roughly half of what it should be.  When I force VGA mode (as opposed to NTSC composite video, which is the default), the games run at 60Hz and the game logic appears to be timed correctly.

It has occurred to me that the other difference between composite video and VGA is that composite video typically has interlacing enabled and VGA does not.  The SPG implementation in WashingtonDC does not take interlacing into account; I need to determine whether the SPG timing is supposed to be effected by interlacing.

**Comment by snickerbockers** [comment]
One important datapoint to keep in mind is that during the spiral-swirl bootup, the vcount is 525 regardless of whether or not interlacing is enabled.  That only makes sense if the SPG skips over lines from the other field when interlacing is enabled; ergo I believe the correct solution to this bug will involve making the SPG increment the line by 2 at a time while also remembering to switch between odd/even lines.

**Comment by snickerbockers** [comment]
I ran the following test program on a real Dreamcast.  I expected to prove that the increments the current scanline by two every pixel-clock cycle in interlaced mode.  What I observed was that it increments by 1, and it goes from 0 to 524 (which is vcount - 1).  I did observe (with a different program) that this video mode does output at 60hz on Dreamcast vs 30hz on WashingtonDC, so it is doubling somewhere but not where I expected it to...

```
#include <kos.h>

#include <stdint.h>

KOS_INIT_FLAGS(INIT_IRQ);

#define SAMPLE_COUNT 1024
static unsigned vpos[SAMPLE_COUNT];
static unsigned n_samples;

static uint32_t volatile * const spg_status = (uint32_t*)0xa05f810c;
static uint32_t volatile * const spg_control = (uint32_t*)0xa05f80d0;
static uint32_t volatile * const spg_load = (uint32_t*)0xa05f80d8;

int main(int argc, char **argv) {
    int x, y;

    /* Set the video mode */
    vid_set_mode(DM_640x480, PM_RGB565);

    if (*spg_control & (1 << 4))
        printf("interlace scan\n");
    else
        printf("progressive scan\n");

    printf("vcount is %u\n", 1 + ((*spg_load >> 16) & 0x3ff));

    uint32_t scanline = *spg_status & 0x3ff;

    while (n_samples < SAMPLE_COUNT) {
        vpos[n_samples++] = scanline;
        do {
            scanline = *spg_status & 0x3ff;
        } while (scanline == vpos[n_samples - 1]);
    }

    unsigned sample_no;
    for (sample_no = 0; sample_no < n_samples; sample_no++)
        printf("%u: %u\n", sample_no, vpos[sample_no]);

    return 0;
}
```

**Comment by snickerbockers** [comment]
This bug has been fixed (sort of) by b61e126040ead00f5da617a93ebc66e09f61d859.
